rfd/0086-stream-audit-events.md

```diff
@@ -0,0 +1,227 @@
+3. [unblock registered watchers](https://github.com/gravitational/teleport/blob/cf205b01a5aa88fd4fcdb499ff9b9c40c4e5c335/lib/backend/dynamo/shards.go#L147)
+4. [receive stream records in a loop](https://github.com/gravitational/teleport/blob/cf205b01a5aa88fd4fcdb499ff9b9c40c4e5c335/lib/backend/dynamo/shards.go#L155)
+	- 4.1. [send stream records to watchers](https://github.com/gravitational/teleport/blob/cf205b01a5aa88fd4fcdb499ff9b9c40c4e5c335/lib/backend/dynamo/shards.go#L170)
+5. [close registered watchers on error](https://github.com/gravitational/teleport/blob/cf205b01a5aa88fd4fcdb499ff9b9c40c4e5c335/lib/backend/dynamo/shards.go#L148)
+
+Of the above, steps 1., 3., 4.1. and 5. are specific for streaming backend changes.
+For this reason, we propose the following:
+- step 1. should be generic so that we can stream any table
+- step 4.1. should be generic so that we can send backend changes to [watchers] and audit events to the interested party
+- steps 3. and 5. are not needed for audit events streaming, so they should be refactored out
+
+[watchers]: https://github.com/gravitational/teleport/blob/cf205b01a5aa88fd4fcdb499ff9b9c40c4e5c335/lib/backend/dynamo/dynamodbbk.go#L550-L553
+
+The above will be achieved by:
+- adding new `StreamConfig` struct that contains the `TableName` from where to stream (for step 1.) and a generic `OnStreamRecords` function (for step 4.1.):
+```go
+type StreamConfig struct {
+	// AwsSession is an AWS session
+	AwsSession *session.Session
+	// PollStreamPeriod is a polling period for event stream
+	PollStreamPeriod time.Duration
+	// TableName is the DynamoDB table that will be streamed
+	TableName string
+	// OnStreamRecords is the function to be called when a stream shard returns a set of records
+	OnStreamRecords func([]*dynamodbstreams.Record) error
+}
+```
+- adding new `NewStream` function that does steps 1. (using `TableName`) and 2.
+- adding new `Poll` function that does step 4. (calling `OnStreamRecords`)
+- removing steps 3. and 5. from `shards.go`
+
+Then, the [`Backend.asyncPollStreams`] function would look something like this:
+```go
+func (b *Backend) asyncPollStreams(ctx context.Context) error {
+	cfg := StreamConfig{
+		AwsSession:       b.session,
+		PollStreamPeriod: b.PollStreamPeriod,
+		TableName:        b.TableName,
+		OnStreamRecords: func(records []*dynamodbstreams.Record) error {
+			events := make([]backend.Event, 0, len(records))
+			for i := range records {
+				event, err := toEvent(records[i])
+				if err != nil {
+					return trace.Wrap(err)
+				}
+				events = append(events, *event)
+			}
+			b.buf.Emit(events...)
+			return nil
+		},
+	}
+
+	for {
+		// there's no stream resuming for backend changes so the stream cursor is empty
+		cursor := ""
+
+		stream, err := NewStream(ctx, cfg) // steps 1. and 2.
+		if err == nil {
+			b.buf.SetInit()                // step 3.
+			err = stream.Poll(ctx, cursor) // steps 4 and 4.1
+			b.buf.Reset()                  // step 5.
```

Comment:
step 5 says "[close registered watchers on error]" but the err check is missing.